File: frontend/src/domain/usecases/LoginUser.ts
import type { IAuthRepository } from "@repositories/IAuthRepository";
import type { AuthToken } from "@entities/AuthToken";
import {
  emailValidator,
  loginPasswordValidator,
} from "@factories/AuthValidatorFactory";
import { MESSAGES } from "@constants/messages";

export class LoginUser {
  private authRepository: IAuthRepository;

  constructor(authRepository: IAuthRepository) {
    this.authRepository = authRepository;
  }

  async execute(email: string, password: string): Promise<AuthToken> {
    if (!email || !password) {
      throw new Error(MESSAGES.AUTH.REQUIRED_FIELDS);
    }

    const emailResult = emailValidator.validate(email);
    if (emailResult.isValid === false) {
      throw new Error(MESSAGES.AUTH.INVALID_EMAIL);
    }

    const passwordResult = loginPasswordValidator.validate(password);
    if (passwordResult.isValid === false) {
      throw new Error(MESSAGES.AUTH.INVALID_PASSWORD);
    }

    return this.authRepository.login(email, password);
  }
}

File: backend/src/domain/usecases/login_user.py
from domain.entities import AuthToken
from domain.repositories import IAuthRepository
from domain.validations import EmailValidation, PasswordValidation


class LoginUser:
    """Use case for user authentication following Clean Architecture."""

    def __init__(self, auth_repository: IAuthRepository):
        self.auth_repository = auth_repository
        self.email_validation = EmailValidation()
        self.password_validation = PasswordValidation()

    async def execute(self, email: str, password: str) -> AuthToken:
        """Execute login use case with validation."""
        if not email or not password:
            raise ValueError("Email and password are required")

        if not self.email_validation.validate(email):
            raise ValueError("Invalid email format")

        # Note: We don't validate password format on login
        # Only validate that credentials match the database
        return await self.auth_repository.login(email, password)
